File: frontend/src/utils/geo.ts
// Vadodara Geolocation Utilities

export const VADODARA_CENTER: [number, number] = [22.3072, 73.1812];

export function haversineKm(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const R = 6371; // Earth radius in km
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLon = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return Number((R * c).toFixed(2));
}

export function etaMinutes(distKm: number, speedKmh = 40): number {
  if (speedKmh <= 0) return 999;
  const minutes = (distKm / speedKmh) * 60;
  return Math.max(1, Math.round(minutes));
}

// Interpolate position between [lat1, lng1] and [lat2, lng2] by fraction t (0 to 1)
export function interpolateGeo(
  start: [number, number],
  end: [number, number],
  fraction: number
): [number, number] {
  const f = Math.max(0, Math.min(1, fraction));
  return [
    Number((start[0] + (end[0] - start[0]) * f).toFixed(6)),
    Number((start[1] + (end[1] - start[1]) * f).toFixed(6)),
  ];
}

File: backend/app/services/geo.py
"""ResQGrid — geo utility functions (pure, no I/O)."""
from __future__ import annotations

import math


def haversine_km(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """Return great-circle distance in kilometres between two WGS-84 points.

    Args:
        lat1: Latitude of point 1 (decimal degrees).
        lng1: Longitude of point 1 (decimal degrees).
        lat2: Latitude of point 2 (decimal degrees).
        lng2: Longitude of point 2 (decimal degrees).

    Returns:
        Distance in kilometres (≥ 0).
    """
    R = 6371.0  # Earth radius in km
    phi1, phi2 = math.radians(lat1), math.radians(lat2)
    dphi = math.radians(lat2 - lat1)
    dlambda = math.radians(lng2 - lng1)
    a = math.sin(dphi / 2) ** 2 + math.cos(phi1) * math.cos(phi2) * math.sin(dlambda / 2) ** 2
    return 2 * R * math.asin(math.sqrt(a))


def eta_minutes(dist_km: float, speed_kmh: float) -> float:
    """Estimated travel time in minutes.

    Args:
        dist_km: Distance in kilometres.
        speed_kmh: Speed in km/h (must be > 0).

    Returns:
        ETA in minutes. Returns 0.0 if distance is 0 or speed is non-positive.
    """
    if speed_kmh <= 0 or dist_km <= 0:
        return 0.0
    return (dist_km / speed_kmh) * 60.0
